core/src/OpenACC/Kokkos_OpenACC_Instance.hpp

```diff
@@ -42,6 +44,20 @@ class OpenACCInternal {
   uint32_t instance_id() const noexcept;
 };
 
+void create_OpenACC_instances(std::vector<OpenACC>& instances);
+
+}  // namespace Kokkos::Experimental::Impl
+
+namespace Kokkos::Experimental::Impl {
+// For each space in partition, assign a new async ID, ignoring weights
+template <class T>
+std::vector<OpenACC> impl_partition_space(const OpenACC& base_instance,
+                                          const std::vector<T>& weights) {
+  std::vector<OpenACC> instances(weights.size());
+  Kokkos::Experimental::Impl::create_OpenACC_instances(instances);
```

Comment:
Minor optimization used in other backends (optional, i.e. just comment that you won't do it and resolve if you don't want to)
```suggestion
  std::vector<OpenACC> instances;
  auto const n = weights.size();
  instances.reserve(n);
  std::generate_n(std::back_inserter(instances), n,
      [] { return OpenACC(OpenACCInternal::m_next_async++); });
```
(not formatted nor tested)

Reiterating that I don't think we need to guarantee thread safety between calls to `partition_space` and that we can just have simplistic scheme to assign new async numbers.
The fanciest I would go would be reserve a range say 123 to 456 and assign them round-robin.
We don't need to care about programs that would create hundreds of these, I don't expect that to be portable.

Reply:
I agree and would prefer a range that is used in a round-robin fashion.

Reply:
Changed the async ID assignment logic to use a simple round-robin assignment over a range.